pkg/apis/coordination/validation/validation.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/kubernetes/pkg/apis/coordination"
+)
+
+// ValidateLease validates a Lease.
+func ValidateLease(lease *coordination.Lease) field.ErrorList {
```

Comment:
Based on prior conversations, I have the following opinions about validation (some of which are already captured in other comments on this PR):
- We should avoid clients having to reason about prior field values.
- We should avoid cross-field validation, especially between critical and debug-only fields (e.g. do not validate RenewTime based on AcquireTime).
- We can still reject obviously incorrect config:
  - negative LeaseDurationSeconds
  - negative LeaseTransitions
  - empty HolderIdentity (as long as this doesn't break legacy applications)
- The extent of additional validation we perform for *transitions* is bounded by how much we can expect from clients:
  - If we can't expect clients to do *any* reasoning about prior values, we shouldn't do extra validation for transitions.
  - But some expected client behavior may be possible to validate:
    - e.g. tight clock skew means clients can probably be expected to make monotonic changes to RenewTime. We should be careful to identify what problems validating this can cause/prevent. If clients typically do this, and controllers just use changes in RenewTime as a trigger to fire a LeaseDurationSeconds timer, rather than reasoning about the RenewTime, validating monotonicity seems like it doesn't add much value (because then nobody cares about monotonicity).
- I'm not sure we should do transition validation for fields that only provide debugging help (AcquireTime, LeaseTransitions), because it sounds like unnecessary complexity to me.
- We should make sure that legacy applications won't depend on validation behavior that differs in this API.

Reply:
I mostly agree with that with two caveats:
1. I think that the fact that field is treated as "debugging help" doesn't really matter - it's in the API anyway.
2. This is basically a new API (and we won't do any automatic transitions). We know what "leaderElection" library in k8s is doing and what it assumes so we know we can migrate it. For other clients, they will need to explicitly switch to use this API, so they will need to understand it anyway, I think.

So I think that doesn't results in any AIs for this PR for now.

Reply:
>So I think that doesn't results in any AIs for this PR for now.

To match my thoughts above, I think we'd need to:
- in ValidateLeaseSpec
  - remove cross-validation of RenewTime and AcquireTime
  - potentially add validation to reject empty HolderIdentity (if this won't break any legacy)
- in ValidateLeaseSpecUpdate
  - @bgrant0607 wanted us to think carefully about whether requiring monotonicity/other special lease-update behavior can cause any issues. Nothing OTTH but I haven't given it a lot of thought...so still possible that we'd remove some of this logic.

> I think that the fact that field is treated as "debugging help" doesn't really matter - it's in the API anyway.

Works for me.

> This is basically a new API (and we won't do any automatic transitions). We know what "leaderElection" library in k8s is doing and what it assumes so we know we can migrate it. For other clients, they will need to explicitly switch to use this API, so they will need to understand it anyway, I think.

SGTM

Reply:
> remove cross-validation of RenewTime and AcquireTime

fair enough - done

> potentially add validation to reject empty HolderIdentity (if this won't break any legacy)

I didn't add that consciously (about legacy - see my comment above).
My usecase for allowing that to be empty is to enable usecase of passing leadership.
E.g. if I'm gracefully terminating, I may want to drop the lease (and we can implement it by setting holder to empty and renew time to 0 or sth like that).

> in ValidateLeaseSpecUpdate

Apparently what I wanted to do in Spec I failed in SpeUpdate - so I remove parts of validation logic.